feat(s3): add to_deltalake_streaming for single-commit Delta writes

skoschik opened the pull request:
## Feature or Bugfix
Feature

## Detail
This PR introduces `wr.s3.to_deltalake_streaming`, a helper that allows users to write **large datasets** to Delta Lake on S3 in a **single atomic commit**.

The existing `wr.s3.to_deltalake` writes one Delta version per call.  
When data must be processed or loaded in chunks (e.g., due to memory limits), this creates multiple versions and complicates time travel.

The new `to_deltalake_streaming` function:
- Accepts an iterable or generator of Pandas DataFrames.
- Streams data through Arrow’s `RecordBatchReader` into one Delta transaction.
- Supports partitioning, schema overwrite/merge, and S3/DynamoDB locking.
- Produces exactly **one Delta version** per run, even for multi-chunk writes.

### Example

```python
def generate_data():
    for i in range(10):
        yield pd.DataFrame({"id": range(i * 1000, (i + 1) * 1000), "value": [i] * 1000})

wr.s3.to_deltalake_streaming(
    dfs=generate_data(),
    path="s3://bucket/delta-table/",
    mode="overwrite",
    s3_allow_unsafe_rename=True
    # lock_dynamodb_table="delta-locks",  # optional
)
```

Result: one Delta commit per rebuild, instead of multiple chunked versions.

## Relates
Enhances Delta Lake integration for S3 single-commit writes.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: f02f89577f8f9cb71b0a512496c95027ed6df296
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=69a3c72e-d401-4756-ab8b-b7021cfb972b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: f02f89577f8f9cb71b0a512496c95027ed6df296
* Result: FAILED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=84c549f5-dd31-4106-9a16-1e66c7afe343%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: ec1bc7ef6a00c235af948bf48f3ac01ed0768af7
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=99219de4-c3cb-433d-aa42-b40245d2057d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubDistributedCodeBuild6-jWcl5DLmvupS
* Commit ID: 530620f658e3f433437639f4d1d7ac9bda70407f
* Result: FAILED
* [Build Logs](https://p0rshq9q5i.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=98cfec6b-df9d-4fa0-ad17-d6a48fb8350b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

kukushking commented:
Thanks @skoschik ! There's some minor linting errors, and would be great to have a test case for the streaming write, otherwise looks great!

skoschik commented:
> Thanks @skoschik ! There's some minor linting errors, and would be great to have a test case for the streaming write, otherwise looks great!

Thanks for the review @kukushking! I've fixed the linting issues and added a few test cases for the streaming write. Let me know if there’s anything else you'd like adjusted.

jaidisido commented:
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-4rfo0GHQ0u9a
* Commit ID: 6a8cc81cbcc71767da1677b937f7aed0cc8a82c0
* Result: SUCCEEDED
* [Build Logs](https://gqgcj9s2gc.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=895296aa-1b44-438d-8dc7-b4fb249f2b56%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
